pkg/ebpf/c/common/buffer.h

```diff
@@ -30,35 +31,46 @@ statfunc buf_t *get_buf(int idx)
 }
 
 // biggest elem to be saved with 'save_to_submit_buf' should be defined here:
-#define MAX_ELEMENT_SIZE bpf_core_type_size(struct sockaddr_un)
+#define MAX_ELEMENT_SIZE sizeof(struct sockaddr_un)
 
 statfunc int save_to_submit_buf(args_buffer_t *buf, void *ptr, u32 size, u8 index)
 {
     // Data saved to submit buf: [index][ ... buffer[size] ... ]
 
-    if (size == 0)
+    barrier();
+    // set argument size bounds
+    if (size == 0 || size > MAX_ELEMENT_SIZE || buf->offset >= ARGS_BUF_SIZE)
         return 0;
 
-    barrier();
-    if (buf->offset > ARGS_BUF_SIZE - 1)
+    u32 buffer_index_offset = buf->offset + 1; // buffer offset after writing the index
+    if (buffer_index_offset > ARGS_BUF_SIZE)
         return 0;
 
-    // Save argument index
+    // Save argument index (add 1 to index later)
     buf->args[buf->offset] = index;
 
-    // Satisfy verifier
-    if (buf->offset > ARGS_BUF_SIZE - (MAX_ELEMENT_SIZE + 1))
+    // Force verifier to compare size register with a bound maximum
+    asm volatile("if %[size] < %[max_size] goto +1;\n"
+                 "%[size] = %[max_size];\n"
+                 :
+                 : [size] "r"(size), [max_size] "i"(MAX_ELEMENT_SIZE));
+
+    u32 buffer_arg_offset = buffer_index_offset + size; // buffer offset after writing the ptr
+
+    // Satisfy verifier - offset+index+size(offset+size+1) must not go above ARGS_BUF_SIZE
+    if (buffer_arg_offset > ARGS_BUF_SIZE)
         return 0;
 
-    // Read into buffer
-    if (bpf_probe_read(&(buf->args[buf->offset + 1]), size, ptr) == 0) {
-        // We update offset only if all writes were successful
-        buf->offset += size + 1;
-        buf->argnum++;
-        return 1;
-    }
+    // Read into buffer after the argument index
+    if (bpf_probe_read_kernel(&(buf->args[buffer_index_offset]), size, ptr) < 0)
```

Comment:
How do you know that we are reading from a kernel pointer here?

Reply:
This is used only in syscall argument extraction which I think is copied from the bpf context first. Anyway we can't be certain and there would be no harm done moving back to the generic probe_read.

Reply:
This actually caused issues in certain syscalls so I reverted.